docs: C2 release creation detailed design for review (#362)
camaraproject/ReleaseManagement: documentation/SupportingDocuments/CAMARA-Release-Creation-Detailed-Design.md

> # Release Creation Process — Detailed Design (MVP)

Reviewer: could be shortened to just "Release Process"

| Issue closure | Not specified | Blocked in SNAPSHOT ACTIVE and DRAFT READY states |
| Tag creation | After Release PR merge | Only at final publication |
| Source of truth | Not specified | `release-metadata.yaml` on snapshot branch |

These refinements are **compatible extensions** to the concept. A change request to update the concept document will follow after design acceptance.

---

**Status:** MVP Design — Recommendations for review.

---

## Table of Contents

1. [User Perspectives and Requirements](#1-user-perspectives-and-requirements)
2. [Design Principles](#2-design-principles)
3. [State Model](#3-state-model)
4. [Detailed Design](#4-detailed-design)
5. [Bot UX Contract](#5-bot-ux-contract)
6. [Permissions and Enforcement](#6-permissions-and-enforcement)
7. [Open Questions](#7-open-questions)
8. [Future Enhancements (Post-MVP)](#8-future-enhancements-post-mvp)

---

## 1. User Perspectives and Requirements

### 1.1 Repository Maintainer

**Role:** Owns the API content and decides when it's ready for release.

**Requirements:**

| ID | Requirement | Rationale | MVP |
|----|-------------|-----------|-----|

> | M1 | Know what release configuration is active | Avoid confusion about versions, statuses | ✅ |

Author: Given the context and the audience I haven't spent the effort for this change.

| R4 | Override capability for stuck releases | Handle exceptional situations |
| R5 | Audit trail of release decisions | Traceability for governance |
| R6 | Know which commit a release is based on | Reproducibility |
| R7 | See current vs. discarded snapshots | Know which attempt is active |

### 1.3 Governance

**Role:** Ensures process integrity, prevents errors, maintains consistency.

**Requirements:**

| ID | Requirement | Rationale |
|----|-------------|-----------|
| G1 | Release configuration comes from `release-plan.yaml` | Single source of truth |
| G2 | Configuration scope frozen during active snapshot | Prevent drift and confusion |
| G3 | Mechanical changes are tamper-proof | Version numbers must be correct |
| G4 | Clear separation: content vs. configuration | Metadata PRs separate from implementation PRs |
| G5 | Immutable release snapshots | What you reviewed is what you release |
| G6 | All releases traceable to source commit | Audit and reproducibility |
| G7 | Consistent process across repositories | Predictable, learnable workflow |

---

## 2. Design Principles

### 2.1 Issue-Based Release Tracking

> **Principle:** Each release (`rX.Y`) is tracked via a dedicated Release Issue (GitHub issue) that serves as the UI, trigger surface, and audit trail across all snapshot attempts.

Reviewer: maybe use "trigger point" rather than "trigger surface" for ease of understanding ?

Author: no, it has more than one trigger point. It's not a user facing document btw.

- Multiple attempts per `rX.Y` are normal and visible
- Defers complexity of forward-merge

### 2.7 Automation Creates, Human Publishes

**Principle:** Automation creates draft releases; humans explicitly approve final publication and tag creation.

**Addresses:** R3, G3

**Rationale:**
- Automation ensures consistency of artifacts
- Human checkpoint prevents accidents
- Clear accountability for publication decision
- Uses GitHub Environment protection
- Tag `rX.Y` only created at final publication (no extensions)

---

## 3. State Model

### 3.1 Key Concepts

**Release (`rX.Y`):** The target release number. Has no extensions. Multiple snapshot attempts may exist for one release.

**Snapshot:** An immutable, automation-owned branch representing one attempt at a release. Identified by `rX.Y-<shortsha>`. Can be discarded and replaced by a new snapshot.

**`release-metadata.yaml`:** The authoritative record for a snapshot, created automatically on the snapshot branch. Contains base commit SHA, snapshot identity, and release parameters. Bot messages derive information from this file.

**Current Snapshot:** The active snapshot for a release. Only one is active at a time.

> **Discarded Snapshot:** A previous snapshot replaced by a newer attempt. Snapshot branch deleted; release-review branch kept for reference.

Reviewer: ```suggestion
**Discarded Snapshot:** A previous snapshot branch replaced by the snapshot branch of the latest release attempt. Snapshot branch deleted; release-review branch kept for reference.
```

> **Discard vs. Revoke:** Before a GitHub draft release exists, snapshots are *discarded*. After a draft exists, the draft is *revoked*, and the automation also discards the underlying snapshot.

Author: Applied, but as I have replaced /revoke-draft with /delete-draft, the text is adapted.

| `release-plan.yaml` | On `main` branch | Release configuration | Persistent, updated per release cycle |
| Release Issue | One per `rX.Y` | UI, trigger surface, audit trail | Created → closed (PUBLISHED or CANCELLED) |
| Snapshot branch | `release-snapshot/rX.Y-<shortsha>` | Mechanical changes (automation-owned) | Created → deleted on discard or after tag |
| `release-metadata.yaml` | On snapshot branch | Authoritative snapshot record | Created with snapshot; source of truth |
| Release-review branch | `release-review/rX.Y-<shortsha>` | Reviewable content (human-owned) | Created → kept until publication (reference for retry) |
| Release PR | PR: release-review → snapshot | CHANGELOG/README/Checklist review | Created → merged or closed |
| Draft release | GitHub Releases | Pre-publication artifact | Created after PR merge |
| Published release | GitHub Releases + tag `rX.Y` | Final artifact | Tag created at publication |

### 4.2 Issue Template

The Release Issue template collects only supplementary information. Technical configuration comes from `release-plan.yaml`.

**Template asks for:**
- Confirmation that `release-plan.yaml` is ready

> - Confirmation that intended PRs are merged

Reviewer: Not sure which PRs are meant here ? the Release PR, or the PR to update README / CHANGELOG (which is also in the next bullet) ?

- Release highlights (for CHANGELOG)
- Additional notes (timing, coordination)

> - Urgency level

Reviewer: this was not addressed before I believe. How do we want this to be expressed ? or can it be part of the previous bullet (e.g. timing/urgency) ?

**Template does NOT ask for:**
- Release type, tag, meta-release (from `release-plan.yaml`)
- API versions, statuses (from `release-plan.yaml`)
- Dependencies (from `release-plan.yaml`)

### 4.3 Snapshot Creation (`/create-snapshot`)

When `/create-snapshot` is run, automation:

1. Checks no snapshot already active for this `rX.Y`
2. Reads `release-plan.yaml` from current HEAD of base branch (`main` or maintenance)
3. Validates schema and consistency
4. Calculates API versions (e.g., `3.2.0-rc.2`)
5. Verifies dependencies exist and are published
6. **If validation fails:** Posts error report, state remains OPEN
7. **If validation passes:**
   - Creates snapshot branch `release-snapshot/rX.Y-<shortsha>`
   - Commits mechanical changes (version replacements, URL updates)
   - Creates `release-metadata.yaml` with base commit SHA and release parameters
   - Creates release-review branch `release-review/rX.Y-<shortsha>`
   - Commits CHANGELOG, README, checklists to release-review branch
   - Opens PR: review → snapshot

>    - Updates label to `release-state: snapshot-active`

Reviewer: ```suggestion
   - Updates the Release Issue label to `release-state: snapshot-active`
```

>    - Posts success comment with links and next steps

Author: applied

### 4.4 Dual-Branch Model

Each snapshot attempt uses two branches:

**1. Snapshot Branch** (automation-owned, immutable)
- Name: `release-snapshot/rX.Y-<shortsha>`
- Created from base branch (`main` or maintenance) at current HEAD
- Contains all mechanical changes
- Contains `release-metadata.yaml` (authoritative snapshot record)
- Never edited after creation
- Protected: only automation can write/delete

**2. Review Branch** (human-owned, editable)
- Name: `release-review/rX.Y-<shortsha>`
- Created from snapshot branch
- Contains reviewable content (CHANGELOG, README, checklists)

> - Merged INTO snapshot branch via PR

Author: applied

  release_tag: r4.1
  release_type: pre-release-rc
  release_date: null                   # Set at publication
  src_commit_sha: abc1234def5678901234567890abcdef12345678  # Set at snapshot creation
  release_notes: "Pre-release for CAMARA Fall26 meta-release."

dependencies:
  commonalities_release: "r3.4 (1.2.0)"
  identity_consent_management_release: "r3.3 (1.1.0)"

apis:
  - api_name: location-verification
    api_version: 3.2.0-rc.2
    api_title: "Location Verification"
  - api_name: location-retrieval
    api_version: 0.5.0-rc.1
    api_title: "Location Retrieval"
```

Bot messages derive all snapshot information from this file.

### 4.6 Content Separation

| Content | Committed To | Editable in PR |
|---------|--------------|----------------|
| `info.version` replacement | Snapshot branch | No |
| Server URL replacement | Snapshot branch | No |
| `x-camara-commonalities` | Snapshot branch | No |
| `release-metadata.yaml` | Snapshot branch | No |
| CHANGELOG section | Release-review branch | Yes |
| README release info | Release-review branch | Yes |
| Readiness checklists | Release-review branch | Yes |

### 4.7 Discard and Retry Flow

```
Issue found during review
       ↓

> Comment: /discard-snapshot API spec bug in location-verification

Reviewer: the format of the string for the message has not been specified I believe e.g. does it need quotes ?

       ↓
Automation:
  - Closes Release PR (not merged)
  - Deletes snapshot branch (release-snapshot/r4.1-abc1234)
  - Keeps release-review branch (release-review/r4.1-abc1234) for reference
  - Updates issue: records discard reason
  - Sets label to `release-state: open`
  - Posts status comment with next steps
       ↓
Fix issues on main via PR(s)
       ↓
Comment: /create-snapshot
       ↓
Automation validates new HEAD (def5678), creates new snapshot
       ↓
New snapshot: release-snapshot/r4.1-def5678
New release-review branch: release-review/r4.1-def5678

> New PR opened

Reviewer: ```suggestion
New Release PR opened
```

       ↓
Maintainer can copy content from discarded release-review/r4.1-abc1234 if needed
```

### 4.8 Revoke Draft Flow

If issues are found after Release PR is merged but before publication:

```
Problem discovered in draft release
       ↓
Comment: /revoke-draft Found critical issue in generated artifacts
       ↓
Automation:
  - Deletes draft release
  - Deletes snapshot branch
  - Keeps release-review branch for reference
  - Sets label to `release-state: open`
  - Posts status comment with next steps
       ↓
Fix issues, then /create-snapshot again
```

> ### 4.9 Snapshot Status Tracking

Reviewer: ```suggestion
### 4.9 Release Status Tracking
```

The Release Issue maintains a snapshot history:

```markdown
## Snapshot History

| Snapshot | Status | Created | Discarded | Reason | Review Branch |
|----------|--------|---------|-----------|--------|---------------|
| `r4.1-def5678` | **Current** | 2026-01-17 | — | — | `release-review/r4.1-def5678` |
| `r4.1-abc1234` | Discarded | 2026-01-15 | 2026-01-16 | API spec bug in location-verification | `release-review/r4.1-abc1234` |
```

Note: Discarded snapshot branches are deleted; only release-review branches are kept for reference.

---

## 5. Bot UX Contract

> The bot acts as a **guided UI**, not just automation glue. Every bot response follows a consistent structure.

Reviewer: ```suggestion
The release process automation is realized through a GitHub bot triggered through commands in the Release Issue. The Release Issue serves as the User Interface (UI) between the maintainers and the automation bot.

The bot acts as a **guided UI**, not just automation glue. Every bot response follows a consistent structure.
```

### 5.1 Mandatory Message Structure

Every bot reaction (command execution, rejection, reopen) must:

**Start with:**

> - Current state (or state transition that occurred)

Reviewer: ```suggestion
- Release state (or state transition that occurred)
```

- Key parameters of that state (derived from `release-metadata.yaml` or repository artifacts)

**End with:**
- A bullet list of **all valid next steps**

### 5.2 State-Specific Parameters

| State | Key Parameters to Show | Source |
|-------|------------------------|--------|
| OPEN | — | — |
| SNAPSHOT ACTIVE | Snapshot ID, base commit SHA, Release PR link | `release-metadata.yaml` |
| DRAFT READY | Draft release link, base commit SHA | `release-metadata.yaml`, GitHub API |

### 5.3 Example Bot Messages

> **After `/create-snapshot` succeeds (OPEN → SNAPSHOT ACTIVE):**

Author: applied

|------------|---------|--------|
| Commonalities | r3.4 | ✅ Published |
| ICM | r3.3 | ✅ Published |

---

**Next steps:**
- Review and merge PR [#456](link-to-pr) to proceed to draft
- `/discard-snapshot <reason>` — discard this snapshot and start over
```

**After `/create-snapshot` fails (OPEN → OPEN):**

```markdown
## ❌ Snapshot Creation Failed

**State:** OPEN (unchanged)

### Validation Errors

- ❌ Dependency `commonalities_release: r3.5` not found (not published yet)
- ❌ API `location-verification` has target_api_status `rc` but test file missing

### Release Configuration (from `release-plan.yaml`)

**Release:** `r4.1` (`pre-release-rc`) — Fall26

| API | Target Version | Status |
|-----|----------------|--------|
| location-verification | `3.2.0-rc` | ❌ Missing tests |
| location-retrieval | `0.5.0-rc` | ✅ OK |

---

**Next steps:**
- Fix validation errors on `main`
- `/create-snapshot` — retry after fixes are merged
```

**Command rejected (wrong state):**

```markdown
## ❌ Command Not Allowed

**Current state:** SNAPSHOT ACTIVE  
**Command:** `/create-snapshot`

A snapshot already exists: `r4.1-abc1234`

---

**Valid actions in this state:**

> - Review and merge PR [#456](link-to-pr) to proceed to draft

Author: applied

1. Reopen the issue manually
2. Replace label `release-state: cancelled` with `release-state: open`
```

---

## 6. Permissions and Enforcement

### 6.1 Command Permissions

| Command | Who May Execute |
|---------|-----------------|
| `/create-snapshot` | Maintainers, Codeowners (write/maintain/admin on repo) |
| `/discard-snapshot` | Maintainers, Codeowners, Release Management team |
| `/revoke-draft` | Maintainers, Codeowners, Release Management team |

Permission is enforced by automation via GitHub API checks (repository permissions or team membership).

### 6.2 Branch Protection

**Snapshot branches (`release-snapshot/*`):**
- Protected
- Writable only by automation
- Deletable only by automation (and admins as break-glass)
- No direct pushes from humans

**Release-review branches (`release-review/*`):**
- Not protected
- Maintainers can push refinements
- Best-effort cleanup by automation after release
- Manual cleanup acceptable

Release-review branches and PRs are best-effort cleanup artifacts. Leaving an old release-review branch or PR is untidy but safe; snapshot branches are the only protected, authoritative artifacts.

### 6.3 PR Labels for Discarded Snapshots

> When a snapshot is discarded, the associated Release PR should be:

Author: That can be done automated as the discard happens as a workflow run. Added "by automation"

- May render a *temporary* draft of calculated versions in the issue comment
- Must never be treated as canonical state

**Rationale:** Allows readiness inspection without side effects. Avoids resurrecting the VALIDATED state implicitly.

### 8.2 Open Pull Request Safeguard (Policy-driven)

A future safeguard may control how open pull requests affect snapshot creation.

**Proposed default behavior (future, not MVP):**
- `/create-snapshot` **blocks** if there are open (non-draft) PRs touching:
  - API definitions
  - `release-plan.yaml`
- Draft PRs **do not block** snapshot creation

**Optional override (explicit intent):**
- `/create-snapshot --allow-open-prs`

**Rationale:** Makes intent explicit when releasing despite ongoing work. Allows teams to encode policy by marking PRs as draft. Keeps MVP lean while supporting stricter future governance.

### 8.3 Explicit Base Branch Selection (Maintenance Releases)

For future support of maintenance releases, the base branch could be specified explicitly.

**Design constraint:**
- The **base branch determines which `release-plan.yaml` applies**
- The issue must not become a state store for the chosen branch

**Proposed UX (future):**

> - Issue template may *ask* for intended base branch (default: `main`) for human coordination

Author: Changed  a bit: ""The Release Issue template may include a field for the intended base branch (default: main if not provided) for human coordination""

| M2 (See calculated versions before triggering) | **Post-MVP:** Preview/Dry-Run command (Section 8.1) |
| M3 (Control start) | `/create-snapshot` is explicit command |
| M4 (Abort and retry) | `/discard-snapshot` + `/create-snapshot` |
| M5 (Refine CHANGELOG) | Release-review branch PR allows edits |
| M6 (Protect mechanical) | Direct commit to snapshot branch, not in PR (by construction) |
| M7 (Clear feedback) | Bot UX contract with actionable messages |
| M8 (Preserve work) | Release-review branch kept on discard |
| M9 (Know current state) | Bot always shows state + valid next steps |
| M10 (Minimal actions) | Single `/create-snapshot` command |
| R1 (Cross-repo visibility) | Issues queryable, labels filterable |
| R2 (Verify dependencies) | Validation checks dependencies exist |
| R3 (Human checkpoint) | Draft + environment approval for publish |
| R4 (Override capability) | RM can run discard/revoke commands |
| R5 (Audit trail) | Issue history, commands, bot responses |
| R6 (Know source commit) | SHA in `release-metadata.yaml` |
| R7 (Current vs. discarded) | Snapshot history table, bot messages |
| G1 (Config from yaml) | Bot shows config from `release-plan.yaml` |

> | G2 (Scoped freeze) | Blocking rule on `release-plan.yaml` PRs for active release |

Reviewer: ```suggestion
| G2 (Scoped freeze) | Blocking rule on PRs for `release-plan.yaml` in active state |
```

Author: Changed to "| G2 (Scoped freeze) | Blocking rule on PRs for `release-plan.yaml` if an active snapshop exists |"

3. **Dual-branch separation**: Each snapshot uses two branches — a protected *snapshot branch* for mechanical changes (versions, URLs) and an editable *review branch* for documentation (CHANGELOG, README). This prevents accidental corruption of version fields by construction.

4. **Repository artifacts as source of truth**: The `release-metadata.yaml` on the snapshot branch is the authoritative record of the release attempt, including base commit SHA. The issue serves as UI, trigger surface, and audit trail — not the state store.

5. **Bot as guided UI**: Every automation response shows current state, key parameters, and all valid next actions — turning the bot into an interactive guide rather than just a log.

### Key Benefits

- **Safe by design**: Mechanical fields protected structurally, not by policy
- **Cheap retries**: Discarding and recreating snapshots is fast and normal
- **Clear accountability**: Explicit commands with audit trail
- **Reduced confusion**: One active snapshot per release, clear state labels
- **Human control**: Automation creates drafts; humans publish

> - **One-step happy path**: Single `/create-snapshot` command validates and creates

Reviewer: ```suggestion
- **One-step happy path**: Single `/create-snapshot` command validates and creates the snapshot branch, the review branch and the review PR (?)
```

### Deviations from Current Concept

| Aspect | Concept Document | This Design |
|--------|------------------|-------------|
| Branch naming | `release/rX.Y` | `release-snapshot/rX.Y-<shortsha>` (SHA-based) |
| Trigger mechanism | Label (`trigger-release`) | Slash command (`/create-snapshot`) |
| Validation | Implicit in trigger | Preflight validation within `/create-snapshot` |

> | PR structure | Single release prep PR | Dual-branch: snapshot (mechanical) + review (documentation) |

Reviewer: in the documentation I propose to continue to use the term Release PR iso Release Preparation PR and I would then expect the issue to be named something like  "Preparation of Release PR (repo-name rX.Y)".

Author: Changed Review PR → Release PR throughout

| Bot responses | Not specified | Mandatory structure with next-steps guidance |
| Issue closure | Not specified | Blocked in SNAPSHOT ACTIVE and DRAFT READY states |

> | Tag creation | After PR merge | Only at final publication |

Reviewer: ```suggestion
| Tag creation | After Release PR merge | Only at final publication |
```

**Note:** M2 (preview before triggering) is addressed by the Preview/Dry-Run extension (Section 8.1). In MVP, calculated versions are shown immediately after `/create-snapshot` succeeds.

### 1.2 Release Manager

**Role:** Coordinates releases across repositories, ensures meta-release alignment.

**Requirements:**

| ID | Requirement | Rationale |
|----|-------------|-----------|
| R1 | Visibility into release status across repos | Coordination for meta-releases |
| R2 | Verify dependencies are published | Can't release if dependencies missing |
| R3 | Human checkpoint before publication | Prevent accidental public releases |
| R4 | Override capability for stuck releases | Handle exceptional situations |
| R5 | Audit trail of release decisions | Traceability for governance |
| R6 | Know which commit a release is based on | Reproducibility |
| R7 | See current vs. discarded snapshots | Know which attempt is active |

### 1.3 Governance

**Role:** Ensures process integrity, prevents errors, maintains consistency.

**Requirements:**

| ID | Requirement | Rationale |
|----|-------------|-----------|
| G1 | Release configuration comes from `release-plan.yaml` | Single source of truth |

> | G2 | Configuration scope frozen during active snapshot | Prevent drift and confusion |

Reviewer: it is not clear where the active state of a snapshot is stored

| G3 | Mechanical changes are tamper-proof | Version numbers must be correct |
| G4 | Clear separation: content vs. configuration | Metadata PRs separate from implementation PRs |
| G5 | Immutable release snapshots | What you reviewed is what you release |
| G6 | All releases traceable to source commit | Audit and reproducibility |
| G7 | Consistent process across repositories | Predictable, learnable workflow |

---

## 2. Design Principles

### 2.1 Issue-Based Release Tracking

> **Principle:** Each release (`rX.Y`) is tracked via a GitHub issue that serves as the UI, trigger surface, and audit trail across all snapshot attempts.

Author: Changed to "dedicated Release Issue (GitHub issue)"

**Addresses:** M1, M9, R1, R5, R7

**Rationale:**
- Provides visible, searchable tracking artifact
- Natural place for discussion and decisions
- Links to PRs, snapshots, and releases
- Preserves history across multiple snapshot attempts
- Shows which snapshot is current vs. discarded
- Familiar GitHub pattern for maintainers

**Note:** The issue is NOT the state store. State is derived from repository artifacts (snapshot branch existence, draft release existence).

### 2.2 Slash Commands for User Intent, Labels for System State

> **Principle:** User actions are expressed via **slash commands** in issue comments. Labels represent **system state only**, not actions.

Author: Updated. 
```
**Principle:** User actions are expressed via **slash commands** in Release Issue comments. Labels on the Release Issue represent **system state only**, not actions. User commands create or remove release-snapshot branches (`release-snapshot/rX.Y-<sha>`) and create release-review branches (`release-review/rX.Y-<sha>`).
```

**Addresses:** M3, M4, M10, R4

**Rationale:**
- **Commands express intent**: What the user wants to do
- **Labels express state**: What state the system is in
- Commands are ephemeral (do not persist as labels)
- Prevents label accumulation and ambiguity
- Commands can be reused
- Clear audit trail of who requested what action

**MVP Command Set:**

| Command | Meaning |
|---------|---------|
| `/create-snapshot` | Validate current HEAD, then create snapshot + review PR |
| `/discard-snapshot <reason>` | Discard the active snapshot attempt |

> | `/revoke-draft <reason>` | Revoke a prepared draft release before publication |

Reviewer: as a non experienced GitHub user, I do not know what revoke implies ? is it really needed in addition to discard ?

**Addresses:** G5, M4, M8

**Rationale:**
- Simple mental model: snapshot is a point-in-time attempt
- `src_commit_sha` always accurate per snapshot (stored in `release-metadata.yaml`)
- Review branch preserved for reference (no lost work)
- Multiple attempts per `rX.Y` are normal and visible
- Defers complexity of forward-merge

### 2.7 Automation Creates, Human Publishes

**Principle:** Automation creates draft releases; humans explicitly approve final publication and tag creation.

**Addresses:** R3, G3

**Rationale:**
- Automation ensures consistency of artifacts
- Human checkpoint prevents accidents
- Clear accountability for publication decision
- Uses GitHub Environment protection
- Tag `rX.Y` only created at final publication (no extensions)

---

## 3. State Model

### 3.1 Key Concepts

**Release (`rX.Y`):** The target release number. Has no extensions. Multiple snapshot attempts may exist for one release.

**Snapshot:** An immutable, automation-owned branch representing one attempt at a release. Identified by `rX.Y-<shortsha>`. Can be discarded and replaced by a new snapshot.

> **`release-metadata.yaml`:** The authoritative record for a snapshot, created on the snapshot branch. Contains base commit SHA, snapshot identity, and release parameters. Bot messages derive information from this file.

Author: Updated

**Current Snapshot:** The active snapshot for a release. Only one is active at a time.

**Discarded Snapshot:** A previous snapshot replaced by a newer attempt. Snapshot branch deleted; review branch kept for reference.

**Discard vs. Revoke:** Before a draft exists, snapshots are *discarded*. After a draft exists, the draft is *revoked*, which also discards the underlying snapshot.

> **Maintenance releases:** For maintenance releases, "main" in this model refers to the relevant maintenance branch (e.g., `maintenance-r3`); states and commands are unchanged.

Reviewer: ```suggestion
**Maintenance releases:** For maintenance releases, the same state model applies but starting from the relevant maintenance branch (e.g., `maintenance-r3`) instead of from `main`; states and commands are unchanged.
```

### 3.2 Release States

> States are mutually exclusive. State is derived from repository artifacts, not stored in the issue.

Reviewer: is labelling the issue not actually a way of storing the state in the issue ?

> States are mutually exclusive. State is derived from repository artifacts, not stored in the issue.

Author: Added: "However, the Release Issue is labelled with a release-state: xxx label reflecting the derived state"

| State | Meaning | Label | Derived From |
|-------|---------|-------|--------------|
| **OPEN** | Issue exists; no active snapshot | `release-state: open` | No `release-snapshot/rX.Y-*` branch exists |
| **SNAPSHOT ACTIVE** | Snapshot branch exists (review PR is a consequence) | `release-state: snapshot-active` | `release-snapshot/rX.Y-*` branch exists |
| **DRAFT READY** | Review PR merged; draft release created | `release-state: draft-ready` | Draft release exists for `rX.Y` |
| **PUBLISHED** | Release published; tag `rX.Y` created | `release-state: published` | Tag `rX.Y` exists; issue closed |
| **CANCELLED** | Issue closed without publication | `release-state: cancelled` | Issue closed while in OPEN state |

**Terminal states:** PUBLISHED and CANCELLED are terminal. Closed issues always have one of these two labels for unambiguous reporting.

**State derivation principle:** Branch and tag existence are the authoritative signals. PRs are consequences of state, not drivers of it.

Labels use a single `release-state:` namespace to ensure mutual exclusivity and enable dashboard aggregation.

### 3.3 Command Set (MVP)

> | Command | Meaning | Allowed States | Resulting State |

Reviewer: ```suggestion
| Command | Meaning | Allowed in State | Resulting State |
```

Author: Updated

|---------|---------|----------------|-----------------|
| `/create-snapshot` | Validate current HEAD, create snapshot + review PR | OPEN only | SNAPSHOT ACTIVE (on success) |
| `/discard-snapshot <reason>` | Discard the active snapshot | SNAPSHOT ACTIVE only | OPEN |
| `/revoke-draft <reason>` | Revoke draft release before publication | DRAFT READY only | OPEN |

**Command semantics:**
- `/create-snapshot` validates **current HEAD at execution time**. If validation fails, no snapshot is created and state remains OPEN. Validation errors are reported in the bot response.
- `/discard-snapshot` and `/revoke-draft` require a reason (for audit trail)
- Commands fail with clear explanation if run in the wrong state

### 3.4 State Transitions

| From | Command / Event | To | Notes |
|------|-----------------|-----|-------|
| — | Create issue | OPEN | |
| OPEN | `/create-snapshot` (success) | SNAPSHOT ACTIVE | Validates HEAD, creates snapshot + review PR |
| OPEN | `/create-snapshot` (failure) | OPEN | Validation failed; errors shown |
| OPEN | Close issue | CANCELLED | Terminal state; no active work lost |
| SNAPSHOT ACTIVE | `/discard-snapshot <reason>` | OPEN | Reason required |

> | SNAPSHOT ACTIVE | Merge review PR | DRAFT READY | |

Author: The Release PR merges release-review INTO release-snapshot. That's clarified within the detailed design, glossary etc and now also in two earlier places (c.f. https://github.com/camaraproject/ReleaseManagement/pull/362/commits/74daef883340af47581921166362e35f3e305f07)

### 3.8 Snapshot Lifecycle

When a snapshot is discarded:
1. Review PR is closed (not merged)
2. **Snapshot branch is deleted** (enables clean detection of active snapshot)
3. Review branch is kept (preserves CHANGELOG work for M8)
4. State returns to OPEN
5. Maintainer can run `/create-snapshot` again

**Discarding snapshots is normal, not failure.** The issue tracks all attempts.

---

## 4. Detailed Design

### 4.1 Artifacts

| Artifact | Naming | Purpose | Lifecycle |
|----------|--------|---------|-----------|
| `release-plan.yaml` | On `main` branch | Release configuration | Persistent, updated per release cycle |
| Release issue | One per `rX.Y` | UI, trigger surface, audit trail | Created → closed (PUBLISHED or CANCELLED) |
| Snapshot branch | `release-snapshot/rX.Y-<shortsha>` | Mechanical changes (automation-owned) | Created → deleted on discard or after tag |
| `release-metadata.yaml` | On snapshot branch | Authoritative snapshot record | Created with snapshot; source of truth |
| Review branch | `review/rX.Y-<shortsha>` | Reviewable content (human-owned) | Created → merged into snapshot or kept on discard |

> | Review PR | PR: review → snapshot | CHANGELOG/README review | Created → merged or closed |

Author: We need to decide if we keep the checklist in the current form, in a reduced form or at all. If it exists and is editable during in the release PR, it should not duplicate information which is covered already in release-metadata.yaml.

| Draft release | GitHub Releases | Pre-publication artifact | Created after PR merge |

> | Published release | GitHub Releases + tag `rX.Y` | Final artifact | Tag created at publication |

Author: Within the MVP of the C2 workflow the publishing of the release is manual (in the GitHub UI). Publishing the draft creates automatically the tag (GitHub functionality).

A /publish-release command can be designed within the C3 workflow design (also the trigger for the post release actions). Out of scope here.
